libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1647,8 +1647,26 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		else if (functionName->name() == "suicide" && functionType->kind() == FunctionType::Kind::Selfdestruct)
 			m_errorReporter.typeError(_functionCall.location(), "\"suicide\" has been deprecated in favour of \"selfdestruct\"");
 	}
-	if (!m_insideEmitStatement && functionType->kind() == FunctionType::Kind::Event)
-		m_errorReporter.typeError(_functionCall.location(), "Event invocations have to be prefixed by \"emit\".");
+
+	// Declaring it here since it's used here and below again
+	bool const abiEncodeV2 = m_scope->sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::ABIEncoderV2);
+
+	if (functionType->kind() == FunctionType::Kind::Event)
+	{
+		if (!m_insideEmitStatement)
+			m_errorReporter.typeError(_functionCall.location(), "Event invocations have to be prefixed by \"emit\".");
+		if (!abiEncodeV2)
+			for (size_t i = 0; i < arguments.size(); ++i)
```

Comment:
We should not check the arguments but rather the parameters, because we first convert to the parameter type and then log that. Of course conversion between structs and anything else should not be possible, but it might in the future.

Reply:
So perhaps this should even be a type check on the event declaration and not on the invocation.

Reply:
Yes! That makes sense